pkg/endpoint/endpoint.go

```diff
@@ -1640,62 +1741,105 @@ type MetadataResolverCB func(ns, podName string) (pod *slim_corev1.Pod, _ []slim
 // either the first successful metadata resolution or when the endpoint is
 // removed.
 //
+// baseLabels contains the list of labels use as "base" for the endpoint.
+// The labels retrieved from 'MetadataResolverCB' will be merged into the
+// baseLabels and put into the endpoint.
+// If this list is empty, the labels set on the endpoint will be
+// replaced by the labels returned 'MetadataResolverCB' as long their source
+// matches the source of the labels already present on the endpoint.
+//
+// restoredEndpoint should be set to 'true' if the endpoint is being restored.
+// If this is set to false and the resolver is unable to retrieve the endpoint
+// labels from k8s, the endpoint will be set with the 'init' identity.
+//
+// blocking - will block this function until the endpoint receives a new
+// security identity, and it is regenerated. If 'false', this
+// operation will be done in the background and 'regenTriggered'
+// will always be 'false'.
+//
 // This assumes that after the initial successful resolution, other mechanisms
 // will handle updates (such as pkg/k8s/watchers informers).
-func (e *Endpoint) RunMetadataResolver(bwm bandwidth.Manager, resolveMetadata MetadataResolverCB) {
-	done := make(chan struct{})
-	controllerName := resolveLabels + "-" + e.GetK8sNamespaceAndPodName()
+func (e *Endpoint) RunMetadataResolver(restoredEndpoint, blocking bool, baseLabels labels.Labels, bwm bandwidth.Manager, resolveMetadata MetadataResolverCB) (regenTriggered bool) {
+	done := make(chan bool)
+	regenTriggeredCh := make(chan bool)
+
+	// make the indexer name
+	e.unconditionalLock()
+	idx := e.resolverIdx
+	e.resolverIdx++
+	e.unlock()
+	controllerName := fmt.Sprintf("%s-%d-%s", resolveLabels, idx, e.GetK8sNamespaceAndPodName())
 	go func() {
 		select {
 		case <-done:
 		case <-e.aliveCtx.Done():
 			return
 		}
+		// we can safely remove the controller since its name is unique. If
+		// it wasn't unique, we would risk of removing the "new" controller if
+		// we executed "RunMetadataResolver" more than once.
 		e.controllers.RemoveController(controllerName)
 	}()
 
 	e.controllers.UpdateController(controllerName,
 		controller.ControllerParams{
 			Group: resolveLabelsControllerGroup,
 			DoFunc: func(ctx context.Context) error {
-				ns, podName := e.GetK8sNamespace(), e.GetK8sPodName()
-				pod, cp, identityLabels, info, _, err := resolveMetadata(ns, podName)
+				regenTriggered, err := e.metadataResolver(ctx, restoredEndpoint, blocking, baseLabels, bwm, resolveMetadata)
```

Comment:
nit: I feel like the refactor here could be split from the commit `pkg/endpoint: keep endpoint labels for their original sources`; The >200 line size of the commit with two potential fixes in there makes it harder to track at a glance whether there are important changes being made as part of this refactor within the bugfix patch, or this is a pure refactor to move the content that was previously directly here in the `DoFunc` declaration into its own function.

Upon a closer look, this *is* making around 150LoC change as part of both refactoring and changing behaviour in the same patch. :(

Reply:
oh yeah. Sorry about that. I tried to reason with doing the refactor on a separate commit but I gave up because it was kind of fixing the bug. IIRC this code didn't exist in the first review which made it even further difficult to split it in seperate commits.

Reply:
Refactoring the function in a first patch like this can drop ~100 lines already, then the functional changes will be smaller on top:

```
diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index e6db5802552b..36176f3a2c96 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -1630,6 +1630,53 @@ func (e *Endpoint) APICanModifyConfig(n models.ConfigurationMap) error {
        return nil
 }

+// metadataResolver will resolve the endpoint's metadata from a metadata
+// resolver.
+//
+//   - bwm - the bandwidth manager used to update the bandwidth policy for this
+//     endpoint.
+//
+//   - resolveMetadata - the metadata resolver that will be used to retrieve this
+//     endpoint's metadata.
+func (e *Endpoint) metadataResolver(ctx context.Context,
+       bwm bandwidth.Manager,
+       resolveMetadata MetadataResolverCB) (err error) {
+       ns, podName := e.GetK8sNamespace(), e.GetK8sPodName()
+       pod, cp, identityLabels, info, _, err := resolveMetadata(ns, podName)
+       if err != nil {
+               e.Logger(resolveLabels).WithError(err).Warning("Unable to fetch kubernetes labels")
+               return err
+       }
+       e.SetPod(pod)
+       e.SetK8sMetadata(cp)
+       e.UpdateNoTrackRules(func(_, _ string) (noTrackPort string, err error) {
+               po, _, _, _, _, err := resolveMetadata(ns, podName)
+               if err != nil {
+                       return "", err
+               }
+               value, _ := annotation.Get(po, annotation.NoTrack, annotation.NoTrackAlias)
+               return value, nil
+       })
+       e.UpdateVisibilityPolicy(func(_, _ string) (proxyVisibility string, err error) {
+               po, _, _, _, _, err := resolveMetadata(ns, podName)
+               if err != nil {
+                       return "", err
+               }
+               value, _ := annotation.Get(po, annotation.ProxyVisibility, annotation.ProxyVisibilityAlias)
+               return value, nil
+       })
+       e.UpdateBandwidthPolicy(bwm, func(ns, podName string) (bandwidthEgress string, err error) {
+               _, _, _, _, annotations, err := resolveMetadata(ns, podName)
+               if err != nil {
+                       return "", err
+               }
+               return annotations[bandwidth.EgressBandwidth], nil
+       })
+       e.UpdateLabels(ctx, identityLabels, info, true)
+
+       return nil
+}
+
 // MetadataResolverCB provides an implementation for resolving the endpoint
 // metadata for an endpoint such as the associated labels and annotations.
 type MetadataResolverCB func(ns, podName string) (pod *slim_corev1.Pod, _ []slim_corev1.ContainerPort, identityLabels labels.Labels, infoLabels labels.Labels, annotations map[string]string, err error)
@@ -1658,38 +1705,10 @@ func (e *Endpoint) RunMetadataResolver(bwm bandwidth.Manager, resolveMetadata Me
                controller.ControllerParams{
                        Group: resolveLabelsControllerGroup,
                        DoFunc: func(ctx context.Context) error {
-                               ns, podName := e.GetK8sNamespace(), e.GetK8sPodName()
-                               pod, cp, identityLabels, info, _, err := resolveMetadata(ns, podName)
+                               err := e.metadataResolver(ctx, bwm, resolveMetadata)
                                if err != nil {
-                                       e.Logger(resolveLabels).WithError(err).Warning("Unable to fetch kubernetes labels")
                                        return err
                                }
-                               e.SetPod(pod)
-                               e.SetK8sMetadata(cp)
-                               e.UpdateNoTrackRules(func(_, _ string) (noTrackPort string, err error) {
-                                       po, _, _, _, _, err := resolveMetadata(ns, podName)
-                                       if err != nil {
-                                               return "", err
-                                       }
-                                       value, _ := annotation.Get(po, annotation.NoTrack, annotation.NoTrackAlias)
-                                       return value, nil
-                               })
-                               e.UpdateVisibilityPolicy(func(_, _ string) (proxyVisibility string, err error) {
-                                       po, _, _, _, _, err := resolveMetadata(ns, podName)
-                                       if err != nil {
-                                               return "", err
-                                       }
-                                       value, _ := annotation.Get(po, annotation.ProxyVisibility, annotation.ProxyVisibilityAlias)
-                                       return value, nil
-                               })
-                               e.UpdateBandwidthPolicy(bwm, func(ns, podName string) (bandwidthEgress string, err error) {
-                                       _, _, _, _, annotations, err := resolveMetadata(ns, podName)
-                                       if err != nil {
-                                               return "", err
-                                       }
-                                       return annotations[bandwidth.EgressBandwidth], nil
-                               })
-                               e.UpdateLabels(ctx, identityLabels, info, true)
                                close(done)
                                return nil
                        },
```

Reply:
For review purposes I just rebased the PR on top of this incremental patch, and it makes that core fix change much more targeted to the changes that are being introduced: https://github.com/joestringer/cilium/tree/pr/fix-cni-add

Reply:
This made it a lot easier to read the changes and check that they all line up. I don't see anything else in this patch after splitting it out. There's still some minor things I'd have a mild preference to further split, but there's few enough changes in a single patch that I can keep it all in my head during review.

Reply:
Feel free to take these or leave them. I've reviewed the rest of the patch at this point.